doc/developer/design/20230411_envelope_upsert_order_by_timestamp.md

```diff
@@ -0,0 +1,94 @@
+- Limiting it to only `TIMESTAMP` because
+    - `OFFSET` is already default. We can make it explicit though if we want.
+    - `PARTITION`, `KEY`, `TOPIC`, would not make a good order by clause.
+    - `HEADERS` by itself would not also make a good order by clause (it’s a json blob), but later we can probably allow expressions so that values could be extracted from it.
+- Limiting to only ascending order because no users have expressed a need for descending.
+
+# Explanation
+
+## Proposal
+
+Make it possible with unsafe-mode to provide custom ORDER BY clause which will override the default ordering by offset. The given ORDER BY identifier should refer to the included TIMESTAMP metadata column.
+
+Examples of valid syntax:
+- `CREATE SOURCE ... INCLUDE TIMESTAMP ENVELOPE UPSERT ( ORDER BY TIMESTAMP )`
+- `CREATE SOURCE ... INCLUDE TIMESTAMP AS ts ENVELOPE UPSERT ( ORDER BY ts )`
+
+Examples of invalid syntax:
+- `CREATE SOURCE ... INCLUDE OFFSET ENVELOPE UPSERT ( ORDER BY TIMESTAMP )` (TIMESTAMP column is not included)
+- `CREATE SOURCE ... INCLUDE PARTITION AS p ENVELOPE UPSERT ( ORDER BY p )` (ORDER BY identifier does not refer to the TIMESTAMP metadata column)
+
+# Reference Explanation
+For envelope upsert, the included metadata columns are appended to the value and eventually persisted. For this feature, we will keep track of the index of the order by field and read the corresponding value to compare when upsert-ing.
+
+Here's an example of the upsert behavior where the source consists of a key, value and the timestamp from metadata and is ordered by the timestamp.
+
+```
+ Previously persisted data        Incoming kafka stream decoded
++------------------------+         +-------------------------+
+| Ok (key1, value1, 100) |         |  Ok (key1, value3, 200) |
+| Ok (key2, value2, 201) |         |  Err(key3, some_error2) |
+| Err(key3, some_error1) |         |  Ok (key2, value4, 200) |
++------------------+-----+         +-----+-------------------+
+                   |                     |
+                   |                     |
+                   |                     |
+                   |                     |
+                   v    Upsert Output    v
+                  +------------------------+
+                  | Ok (key1, value3, 200) |
+                  | Ok (key2, value2, 201) |
+                  | Err(key3, some_error2) |
+                  +------------------------+
+```
+Note: As shown in the example above, the errors are still implicitly ordered by offset as we do not persist any extra metadata for them separately. A later error with the same key will always overwrite a previous one.
+
+# Rollout
+This feature will be rolled out behind the unsafe flag, then upgraded to only being available behind a LaunchDarkly flag.
+
+## Testing and Observability
+
+The custom ordering can be tested via testdrive tests in the CI.
+
+## Lifecycle
+
+What will it take to promote this to the next stage i.e. Alpha?
```

Comment:
One thing to watch out for is that as soon as we roll this out to any customers, the syntax will be semi-stable, in that we'll have to be willing to write a migration if we want to change it, and also warn all the customers who were using the old syntax that the syntax has changed.

Reply:
This is only true after its not `unsafe`, right?